cfgmgr/buffermgrdyn.cpp

```diff
@@ -0,0 +1,1620 @@
+
+    SWSS_LOG_DEBUG("processing command:%s PORT table key %s", op.c_str(), port.c_str());
+
+    port_info_t &portInfo = m_portInfoLookup[port];
+
+    SWSS_LOG_DEBUG("Port Info for %s before handling %s %s %s",
+                   port.c_str(),
+                   portInfo.speed.c_str(), portInfo.cable_length.c_str(), portInfo.gearbox_model.c_str());
+
+    task_process_status task_status = task_process_status::task_success;
+
+    if (op == SET_COMMAND)
+    {
+        for (auto i : kfvFieldsValues(t))
+        {
+            if (fvField(i) == "speed")
+            {
+                speed_updated = true;
+                portInfo.speed = fvValue(i);
+            }
+            else if (fvField(i) == "mtu")
+            {
+                mtu_updated = true;
+                portInfo.mtu = fvValue(i);
+            }
+            else if (fvField(i) == "admin_status")
+            {
+                admin_status_updated = true;
+            }
+        }
+
+        string &cable_length = portInfo.cable_length;
+        string &mtu = portInfo.mtu;
+        string &speed = portInfo.speed;
+
+        if (cable_length.empty() || speed.empty())
+        {
+            SWSS_LOG_WARN("Cable length for %s hasn't configured yet, unable to calculate headroom", port.c_str());
+            // We don't retry here because it doesn't make sense until the cable length is configured.
+            return task_process_status::task_success;
+        }
+
+        if (speed_updated || mtu_updated)
+        {
+            SWSS_LOG_INFO("Updating BUFFER_PG for port %s due to speed or port updated", port.c_str());
+
+            //Try updating the buffer information
+            switch (portInfo.state)
+            {
+            case PORT_INITIALIZING:
+                portInfo.state = PORT_READY;
+                if (mtu.empty())
+                {
+                    // During initialization, the flow can be
+                    // 1. mtu, cable_length, speed
+                    // 2. cable_length, speed, mtu
+                    // 3. cable_length, speed, but no mtu specified
+                    // it's ok for the 1st case
+                    // but for the 2nd case, we can't wait mtu for calculating the headrooms
+                    // because we can't distinguish case 2 from case 3 which is also legal.
+                    // So once we have speed updated, let's try to calculate the headroom with default mtu.
+                    // The cost is that if the mtu is provided in the following iteration
+                    // the current calculation is of no value and will be replaced.
+                    mtu = DEFAULT_MTU_STR;
+                }
+                task_status = refreshPriorityGroupsForPort(port, speed, cable_length, mtu);
+                break;
+
+            case PORT_READY:
+                task_status = refreshPriorityGroupsForPort(port, speed, cable_length, mtu);
+                break;
```

Comment:
For both port_table and cable_len table,we could invoke refreshPriorityGroupsForPort only when the port is in PORT_READY state. We could remove refresh in port_init state. Seems to be redundant.

Reply:
Take `handlePortTable` as an example:
Reaching `case PORT_INITIALIZING` means:
- the current state of the port is `PORT_INITIALIZING` because it lacks some info
- an event that supplies the lacking info, `speed` in this case, is being handled and the port's state is about to be transmitted to `PORT_READY`.

Hence, the `refreshPriorityGroupsForPort` should be called for the "first" time once its state is `PORT_READY`.